server/aesdsocket.c

```diff
@@ -9,169 +9,317 @@
+    graceful_shutdown();
+
+    pthread_join(timer_thread, NULL);
+    pthread_mutex_destroy(&g_mutex);
+    closelog();
+    return 0;
+}
+
 
-    if (bind(sockfd, servinfo->ai_addr, servinfo->ai_addrlen) == -1) 
+void* client_thread_func(void* thread_param)
+{
+    client_thread_t *tinfo = (client_thread_t*)thread_param;
+
+    char *ip_str = inet_ntoa(tinfo->client_addr.sin_addr);
+    syslog(LOG_INFO, "Accepted connection from %s", ip_str);
+
+    char buffer[BUFFER_SIZE];
+    ssize_t bytes_received;
+    
+    while(!g_exit_flag)
     {
-        syslog(LOG_ERR, "bind failed");
-        freeaddrinfo(servinfo);
-        close(sockfd);
-        return -1;
+        bytes_received = recv(tinfo->client_fd, buffer, BUFFER_SIZE, 0);
+        if (bytes_received <= 0) 
+        {
+            break;
+        }
+        
+        pthread_mutex_lock(&g_mutex);
+        FILE *file = fopen(FILE_PATH, "a+");
+        if(!file) 
+        {
+            syslog(LOG_ERR, "Failed to open %s for append", FILE_PATH);
+            pthread_mutex_unlock(&g_mutex);
+            break;
+        }
+        fwrite(buffer, 1, bytes_received, file);
+        fflush(file);
+
+        bool found_newline = (memchr(buffer, '\n', bytes_received) != NULL);
+        if(found_newline) 
+        {
+            rewind(file);
+
+            size_t nread;
+            while((nread = fread(buffer, 1, BUFFER_SIZE, file)) > 0) 
+            {
+                send(tinfo->client_fd, buffer, nread, 0);
+            }
+        }
+        fclose(file);
+        pthread_mutex_unlock(&g_mutex);
     }
-    freeaddrinfo(servinfo);
 
-    if (daemon_mode) 
+    shutdown(tinfo->client_fd, SHUT_RDWR);
+    close(tinfo->client_fd);
+
+    syslog(LOG_INFO, "Closed connection from %s", ip_str);
+
+    pthread_mutex_lock(&g_mutex); 
+    SLIST_REMOVE(&g_thread_list_head, tinfo, client_thread_s, entries);
+    pthread_mutex_unlock(&g_mutex);
+
+    free(tinfo);
+    return NULL;
+}
+
+void* timestamp_thread_func(void* arg)
+{
+    while(!g_exit_flag)
     {
-        daemonize();
+        sleep(TIMESTAMP_INTSEC);
+        if(g_exit_flag) break;
+
+        time_t now = time(NULL);
+        struct tm* tmp = localtime(&now);
+        if(!tmp) continue;
+
+        char timestr[128];
+        strftime(timestr, sizeof(timestr), "timestamp:%a, %d %b %Y %T %z\n", tmp);
+
+        pthread_mutex_lock(&g_mutex);
+        FILE* file = fopen(FILE_PATH, "a");
+        if(file) 
+        {
+            fputs(timestr, file);
+            fclose(file);
+        }
+        pthread_mutex_unlock(&g_mutex);
     }
+    return NULL;
+}
+
 
-    if (listen(sockfd, SOMAXCONN) == -1) 
+void graceful_shutdown(void)
+{
+    if(g_server_socket != -1) 
     {
-        syslog(LOG_ERR, "listen failed");
-        close(sockfd);
+        close(g_server_socket);
+        g_server_socket = -1;
+    }
+
+    client_thread_t *tnode = NULL;
+    while(!SLIST_EMPTY(&g_thread_list_head))
```

Comment:
I don't really understand how this is going to work... I don't see where the removal is occurring in this loop.  I think pthread_join is missing in your cleanup based on the valgrind warning.